feat(create-vite): add help usage

edvardsanta opened the pull request:
### Description



<!-- What is this PR solving? Write a clear description or reference the issues it solves (e.g. `fixes #123`). What other alternatives have you explored? Are there any parts you think require more attention from reviewers? -->
closes #16374 
![image](https://github.com/vitejs/vite/assets/75655486/b5790331-bacf-48c0-9828-28737bc06c4e)


<!----------------------------------------------------------------------
Before creating the pull request, please make sure you do the following:

- Read the Contributing Guidelines at https://github.com/vitejs/vite/blob/main/CONTRIBUTING.md.
- Check that there isn't already a PR that solves the problem the same way. If you find a duplicate, please help us reviewing it.
- Update the corresponding documentation if needed.
- Include relevant tests that fail without this PR but pass with it.

Thank you for contributing to Vite!
----------------------------------------------------------------------->

edvardsanta committed:
Create help argument

- help argument will be used as "--help" and by default it is false

edvardsanta committed:
Create function to display help

- i want to find a way to decouple the text from the code

edvardsanta committed:
Modified display

- Using alias to define template and help

bluwy commented on packages/create-vite/src/index.ts:
```diff
@@ -565,6 +575,34 @@ function editFile(file: string, callback: (content: string) => string) {
+function getUsageInfo() {
+  const formattedFrameworkNames = getFormattedVariantNames()
+  return `Usage: create-vite [OPTION]... [DIRECTORY]
+Create a new JavaScript project based on Vite.
```
nit
```suggestion
Create a new Vite project in JavaScript or TypeScript.
```

bluwy commented on packages/create-vite/src/index.ts:
```diff
@@ -565,6 +575,34 @@ function editFile(file: string, callback: (content: string) => string) {
+${formattedFrameworkNames.join('\n')}`
+}
+
+function getFormattedVariantNames() {
+  return FRAMEWORKS.filter(
+    (framework) => framework.name !== 'others' && framework.variants.length > 0,
+  ).map((framework) => {
+    const variantNames = framework.variants.map((variant) => variant.name)
+    return `${variantNames.join(' ')}`
+  })
+}
```
I think we should also put the variants each on their new line as well, otherwise it may be hard to read if some are on a single line.

Also, I think we should filter out the templates that have `customCommand` as they're redirects to other CLIs.

edvardsanta replied:
> I think we should also put the variants each on their new line as well, otherwise it may be hard to read if some are on a single line.
> 
> Also, I think we should filter out the templates that have `customCommand` as they're redirects to other CLIs.

```

Usage: create-vite [OPTION]... [DIRECTORY]
Create a new Vite project in JavaScript or TypeScript.

With no arguments, start an interactive command line dialog.

Options:
  -t, --template NAME        use this template

Available templates:
vanilla-ts
vanilla
vue-ts
vue
react-ts
react-swc-ts
react
react-swc
preact-ts
preact
lit-ts
lit
svelte-ts
svelte
solid-ts
solid
qwik-ts
qwik

```

bluwy reviewed (commented):
Some formatting nits, otherwise a nice addition to me!

edvardsanta committed:
Decouple help message for improved maintainability

- To improve maintainability, the help message has been moved to a separate file.
- Implemented the suggestions provided in the pull request.

edvardsanta committed:
modified test

edvardsanta committed:
Merge branch 'main' into feat/add-help-create-vite

edvardsanta committed:
Merge branch 'main' into feat/add-help-create-vite

bluwy committed:
chore: hardcode the help message

bluwy commented:
Pushed a commit with some styling changes after discussing with other maintainers, here's how it looks now: 
<img width="441" alt="image" src="https://github.com/vitejs/vite/assets/34116392/0308f92c-0101-4228-8611-e3f37220f047">

The hardcoding isn't the most ideal, but I think it simplifies the code a lot and the templates aren't something we update often.

edvardsanta commented:
> Pushed a commit with some styling changes after discussing with other maintainers, here's how it looks now: <img alt="image" width="441" src="https://private-user-images.githubusercontent.com/34116392/339409391-0308f92c-0101-4228-8611-e3f37220f047.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTgyOTIxNDQsIm5iZiI6MTcxODI5MTg0NCwicGF0aCI6Ii8zNDExNjM5Mi8zMzk0MDkzOTEtMDMwOGY5MmMtMDEwMS00MjI4LTg2MTEtZTNmMzcyMjBmMDQ3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA2MTMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNjEzVDE1MTcyNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTgzNDQ0MzQyYTQ0MDRlZjM4MGYyMDlkOTI1ODhhMTY1YjViYjYwODExOTc5MDNjMzRhNmE3NDU5OTJhMWRlNjImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.7NWn7LB0QAlM4wYcAqbak8iSXw6C9vCTIPVgV-VdL9w">
> 
> The hardcoding isn't the most ideal, but I think it simplifies the code a lot and the templates aren't something we update often.

Wouldn't it be interesting to use the color properties in the Framework type? It would make the process more dynamic instead of manually creating them for each new framework.

bluwy commented:
Yeah I did try that but to do so and others, it slows creeps in as a lot of code compared to hardcoding. So I leaned directly to the simplest way for now.

edvardsanta commented:
> Yeah I did try that but to do so and others, it slows creeps in as a lot of code compared to hardcoding. So I leaned directly to the simplest way for now.

I tried to make it more dynamic, but I forgot that we have four types of React, and the variants ended up looking strange.
![image](https://github.com/vitejs/vite/assets/75655486/0c3e5993-4aa3-4717-9b0c-53a88a2e1723)

bluwy commented:
Yeah that was also the part where I threw my hand and went for the simpler path instead 😄 It's hard to sort that one with a specific hardcoding too.
